tests/fv/flannel_migration_test.go

```diff
@@ -0,0 +1,381 @@
+		})
+	})
+
+	Context("Should migrate FV tests", func() {
+		AfterEach(func() {
+			migrationController.Stop()
+		})
+
+		It("Should report nothing to do if Flannel daemonset not exists", func() {
+			// remove flannel daemonset
+			_, err := apiserver.ExecOutput("kubectl", "delete", "daemonset", flannelDs, "-n", "kube-system")
+			Expect(err).ShouldNot(HaveOccurred())
+
+			startController()
+
+			w := migrationController.WatchStderrFor(regexp.MustCompile(`.*no migration process is needed.*`))
+			Eventually(w, "10s").Should(BeClosed(),
+				"Timed out waiting for migration controller report 'no migration process is needed'")
+		})
+
+		It("Should report error if Flannel daemonset has addon manager label", func() {
+			// add addon manager label
+			_, err := apiserver.ExecOutput("kubectl", "label", "daemonset", flannelDs, "-n", "kube-system", "addonmanager.kubernetes.io/mode=EnsureExists")
+			Expect(err).ShouldNot(HaveOccurred())
+
+			startController()
+
+			w := migrationController.WatchStderrFor(regexp.MustCompile(`.*abort migration process.*`))
+			Eventually(w, "10s").Should(BeClosed(),
+				"Timed out waiting for migration controller report 'abort migration process'")
+		})
+	})
+
+	Context("IPAM migrate FV tests", func() {
+		checkCalicoIPAM := func() {
+			// Wait for ipam migration is done.
+			w := migrationController.WatchStderrFor(regexp.MustCompile(`.*nodes completed IPAM migration process.*`))
+			Eventually(w, "10s").Should(BeClosed(),
+				"Timed out waiting for migration controller report 'nodes completed IPAM migration process'")
+
+			Expect(len(flannelCluster.FlannelNodes)).To(Equal(3))
+			// Check Calico IPAM.
+			validateCalicoIPAM(flannelCluster, calicoClient, bc)
+		}
+
+		AfterEach(func() {
+			stopController()
+		})
+
+		It("Should create ippool, felixconfiguration and block affinities", func() {
+			startController()
+			checkCalicoIPAM()
+		})
+
+		It("Should support controller restart after IPAM is done", func() {
+			startController()
+			checkCalicoIPAM()
+
+			stopController()
+
+			startController()
+			checkCalicoIPAM()
+		})
+
+		It("Should support Canal with a default ippool and vxlan disabled", func() {
+			p := api.NewIPPool()
+			p.Name = "default-ipv4-ippool"
+			p.Spec.CIDR = "192.168.0.0/16"
+			p.Spec.BlockSize = 26
+			p.Spec.NodeSelector = "all()"
+			p.Spec.Disabled = false
+			_, err := calicoClient.IPPools().Create(context.Background(), p, options.SetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			// Remove flannel daemonset, add Canal daemonet.
+			_, err = apiserver.ExecOutput("kubectl", "delete", "daemonset", flannelDs, "-n", "kube-system")
+			Expect(err).ShouldNot(HaveOccurred())
+			flannelCluster.AddCanalDaemonset("canal")
+
+			startController()
+			checkCalicoIPAM()
+		})
+	})
+
+	Context("Node ordering FV tests", func() {
+		checkNodeOrdering := func() {
+			// Set watch for node index.
+			w0 := migrationController.WatchStderrFor(regexp.MustCompile(`.*node-2\[index 0\].*`))
```

Comment:
Is this the best way to tell when the migration controller has finished a node?

Can't we check for the labels on that node to know when it was completed? I'm just a bit skeptical of tying our tests to specific wording in our log messages.

Reply:
Ah, I guess it's doing more than just making sure they all complete - it is also checking that the node matches the index. I guess this is OK.

Reply:
Yes. It is about node been processed in correct order.